### PR TITLE
fix github pages deploy guide

### LINKED ISSUE #1355: Flesh out the GitHub Pages deployment guide
Hi! I followed the [github deployment guide](https://docs.astro.build/en/guides/deploy/github/) and for me it just flat out doesn't work.

## 1. Base

The guide states:

> Set the **site** and, if needed, **base** options in astro.config.mjs.
> **base** should be your repository’s name starting with a forward slash, for example /my-repo
> If your repository is named <YOUR_USERNAME>.github.io, you don’t need to include base.

My repository is called "Web" so I set base to "/Web" and the deployment completely broke my website. While assets were in the /assets folder, pages were looking for them in /Web/assets.

The solution was to not include any base parameter in astro.config.mjs.
Why? My website and repository are not called <YOUR_USERNAME>.github.io, so why do I have to omit the base parameter, and not set it like the guide says?

## 2. Deployment source

The guide states:

> On GitHub, go to your repository’s Settings tab and find the Pages section of the settings.
> Choose the gh-pages branch and the "/" (root) folder as the Source of your site and press Save.

I did and the deployment action failed with the error `Invalid deployment branch and no branch protection rules set in the environment`.

Why? How does one set "branch protection rules"? Maybe that it could be included in the guide?

The solution was to set the source to "GitHub Actions" instead of "Deploy from a branch" altogether. It might be in Beta, but it works without any issues.

image

---

I managed to figure it out but it wasn't exactly a smooth experience.
I think it would be great to improve it so other people don't run into the same issues.

### PR DESCRIPTION
**Pull request by kevinzunigacuellar**
#### What kind of changes does this PR include?

- New or updated content

#### Description [(demo)](https://github.com/kevinzunigacuellar/gh-pages-rep)

- Closes #1355 and #1376
- Reorder steps: Pushing and committing the workflow before setting up pages in the repository throws an [error](https://github.com/kevinzunigacuellar/gh-pages-rep/runs/7964873732?check_suite_focus=true#step:2:27)

  ```shell
  Failed to create deployment (status: 404) with build version 
  fcb3c6ad752990f1ea041c7971c0c3f71ac43117. Ensure GitHub Pages has been enabled.
  ```
- Switch **Source** from branch to Github Actions: deploying from branch uses Jekyll to deploy [(error link)](https://github.com/kevinzunigacuellar/gh-pages-rep/runs/7964905392?check_suite_focus=true)

- Add a caution aside for people that added base only to deploy to GitHub pages to update their routes relative to `base`.

- Added a fancy codeblock title

### CODE REVIEW

**Review by fredoist — APPROVED** [review]
Good work! 🔥

**Comment by swift502** [comment]
These changes are great so far! Though maybe I'm alone, but I still don't understand this sentence:

> If your repository is named <YOUR_USERNAME>.github.io, you don’t need to include base.

My website is named "jblaha.art". My repo is called "Web". So I thought I had to include the base parameter. But I didn't. It works without it. What is the purpose of this note?

**Comment by sarah11918** [comment]
Hey @kevinzunigacuellar! Thanks for this! As you know, Chris and I are only checking in periodically on PRs this week, so things are moving a little slower here. 

I just wanted to make sure you noticed that there are two other PRs/Issues (can't exactly remember!) about this very page... You probably already did, but since I'm not able to look closely right now, can you please look for them and try to amalgamate all the various bits people have called out or suggested as helpful for this page? Maybe tag those community members specifically in here so we can omg finally stop working on this page!!  (Just kidding... I love all my Docs pages equally....) 💜

**Comment by kevinzunigacuellar** [comment]
> These changes are great so far! Though maybe I'm alone, but I still don't understand this sentence:
> 
> > If your repository is named <YOUR_USERNAME>.github.io, you don’t need to include base.
> 
> My website is named "jblaha.art". My repo is called "Web". So I thought I had to include the base parameter. But I didn't. It works without it. Actually, it's much better without it, because I don't have to update all my routes.
> 
> What is the purpose of this note?

I think this configuration is for when the name of your repo is exactly the same as your account. Yeah I didn't get it right of the bat as well, It could use a rewording.

**Comment by swift502** [comment]
@kevinzunigacuellar even then I don't understand how the base would affect the build. In my case the base parameter just offset static assets by the base path, but it ultimately seemed to have no effect on how the project was deployed.

**Comment by kevinzunigacuellar** [comment]
Base is meant exactly for that, to shift the built files by a base path. It's very convenient for github pages since they have a per repo deployment.

**Comment by swift502** [comment]
@kevinzunigacuellar Could we then come up with an example of when it is useful, and put it on the page to make it clear?
And also, should it be done by default? As far as I can tell, if you don't already need a base path, you don't need to add one just so you can deploy. (It works for me that way) But the guide currently states it should be there by default.

**Comment by kevinzunigacuellar** [comment]
I think it's possible to skip the base path to deploy but it can break other deployments if the user has multiple matching paths. In my opinion, we should try to follow Github's defaults and try to have all the page assets under the repository's name (e.g `bob.github.io/repo-1` will contain all assets under `/repo-1` and `bob.github.io/repo-2` will contain all assets under `/repo-2`) so we may avoid some conflicts.

The note is there for the people that want to deploy in their root domain `username.github.io` using their matching username repo.

File: src/pages/en/guides/deploy/github.md
Comment (kevinzunigacuellar) [inline src/pages/en/guides/deploy/github.md:21]:
```suggestion
    If your repository has the same name as your username it will be deployed in `<YOUR_USERNAME>.github.io`, hence, you don’t need to include a `base`.
```

Comment (kevinzunigacuellar) [reply src/pages/en/guides/deploy/github.md:21]: @swift502 what do you think about this reword?

Comment (fredoist) [reply src/pages/en/guides/deploy/github.md:21]: Shouldn't the same apply for custom domains?

Comment (swift502) [reply src/pages/en/guides/deploy/github.md:21]: @kevinzunigacuellar Ohhhkay, I think I get it. Thanks for being patient with me.

@fredoist is right. You must _not_ include the base if you're using a custom domain, otherwise assets fail to load. I just tested that on my website.

I'd be much more explicit, because it's so convoluted:

> If you're using GitHub's default site url, e.g. `<YOUR_USERNAME>.github.io/my-repo`, set the `base` parameter to your repository’s name starting with a forward slash, for example `/my-repo`. This is so that Astro understands your website's root is "/my-repo", rather than default "/".
> 
> __Don't__ set the `base` parameter if:
> 
> - your repository itself is called `<YOUR_USERNAME>.github.io`
> - you're using a custom domain

---

Combined with the "update your routes" caution box you added, it feels appropriately comprehensive.
Maybe it's a lot of text though. I'll leave it up to you.

Btw:
> If your repository has the same name as your username

I'm not sure that's a good wording because it implies swift502/swift502, which would be the special user repository, not swift502.github.io right?

Comment (kevinzunigacuellar) [reply src/pages/en/guides/deploy/github.md]: Yes, my mistake here 😅. I decided to keep the part of the original message and add a little more explanation for why and when you need to set a base.

File: src/pages/en/guides/deploy/github.md
Comment (kevinzunigacuellar) [inline src/pages/en/guides/deploy/github.md:23]:
```suggestion
    - your repository itself is named <YOUR_USERNAME>.github.io
```

File: src/pages/en/guides/deploy/github.md
Comment (fredoist) [reply src/pages/en/guides/deploy/github.md]: LGTM 👍🏼

Comment (swift502) [reply src/pages/en/guides/deploy/github.md]: Yeah I like it.

**Review by swift502 — APPROVED** [review]

**Comment by sarah11918** [comment]
@kevinzunigacuellar Let me know when you all are happy with this, then @Yan-Thomas or I will take a quick look over it and can merge!

**Comment by kevinzunigacuellar** [comment]
> @kevinzunigacuellar Let me know when you all are happy with this, then @Yan-Thomas or I will take a quick look over it and can merge!

we are ready for review 😇

File: src/pages/en/guides/deploy/github.md
Comment (Yan-Thomas) [inline src/pages/en/guides/deploy/github.md:24]: We tend to go with the AP Stylebook recommendation for bullet points, I would change this for consistency:
```suggestion
    - Your repository is named `<YOUR_USERNAME>.github.io`.
    - You're using a custom domain.
```

**Review by Yan-Thomas — COMMENTED** [review]
Amazing job @kevinzunigacuellar, I just have a small nitpick but after this LGTM! Love to see this many contributors in a PR 💜

File: src/pages/en/guides/deploy/github.md
Comment (sarah11918) [inline src/pages/en/guides/deploy/github.md:29]:
```suggestion
        If you did not previously have a value for `base` set, and are only configuring this value so that you can deploy to GitHub, you must update your internal page links to now include your `base`.
```
What do you (all!) think about something more "positively" phrased (you must do this), rather than "Don't forget  to!" which may or may not have been clear to the reader before?

File: src/pages/en/guides/deploy/github.md
Comment (sarah11918) [inline src/pages/en/guides/deploy/github.md:19]: Is it too much to show an `astro.config.mjs` example here with those fields filled out?

File: src/pages/en/guides/deploy/github.md
Comment (sarah11918) [inline src/pages/en/guides/deploy/github.md:17]: I know this was existing text when you got here, but "should be something like" without more explicit instruction makes me nervous. 😅 

I think I recall from earlier versions of this page that you SHOULD be using `https://<YOUR_USERNAME>.github.io` -- but, is it possible to deploy TWO DIFFERENT sites on GitHub pages? What would you do then, since you only have one username.

I think it might also be helpful to be more DIRECT about saying to use this for `site` (like, just use this!) or to add what another "something like" could be, for example in the case of having two sites deployed. (Is that even possible?)

**Review by sarah11918 — COMMENTED** [review]
Great work, everyone! 🥳  Not gonna lie, this page has caused me a lot of pain and suffering for a LONG time... thank you for jumping in and working so carefully to get it right!

I left some small comments! See what you think about them, and ping me when you've decided!

File: src/pages/en/guides/deploy/github.md
Comment (swift502) [reply src/pages/en/guides/deploy/github.md:17]: Multiple sites are possible, and as far as I can tell, the `site` for them is the same. `sarah11918.github.io/blog` and `sarah11918.github.io/portfolio` both share a site of `sarah11918.github.io`.

The other "something like" is when you use a custom domain. Then it's the custom domain you fill in as `site`.

File: src/pages/en/guides/deploy/github.md
Comment (swift502) [reply src/pages/en/guides/deploy/github.md:19]: @odwrotnie was asking for an example in #1376. They managed to [misplace the parameters](https://github.com/instap-app/website-definition/blob/4fd94706505c385128284a3f0b27eb85ff2d85ca/astro.config.mjs#L9-L10) outside of the defineConfig object, because they didn't find this clear enough.

File: src/pages/en/guides/deploy/github.md
Comment (sarah11918) [reply src/pages/en/guides/deploy/github.md:17]: Ah, good call, that's where base comes in !

I thought about also suggesting that we show a custom domain as a possiblility there, but then our tip for "Want to use a custom domain?" is alllll the way down at the bottom. And, THAT doesn't say to update SITE, so I didn't know whether that cname file would actually just "overwrite" a github.io site, or whether you were supposed to update SITE, too.

So then, what about something like:

```
- `site` should be `https://<YOUR_USERNAME>.github.io` or `https://my-custom-domain.com`
```

Comment (kevinzunigacuellar) [reply src/pages/en/guides/deploy/github.md:17]: You are right, we forgot to include an update site in that little tip 😅. my best guess is that this file sets an alias for `username.github.io` but the build is using site, whoops 😇

Comment (kevinzunigacuellar) [reply src/pages/en/guides/deploy/github.md:17]:
```suggestion
    - `site` be should be `https://<YOUR_USERNAME>.github.io` or `https://my-custom-domain.com`
```

File: src/pages/en/guides/deploy/github.md
Comment (sarah11918) [inline src/pages/en/guides/deploy/github.md:17]:
``````suggestion
    
```astro title="astro.config.mjs"
import { defineConfig } from 'astro/config'

export default defineConfig({
  site: 'https://astronaut.github.io'
  base: '/my-repo'
})
```
    - `site` should be `https://<YOUR_USERNAME>.github.io` or `https://my-custom-domain.com`
``````

Trying this here to see how it looks *before* the explanation.

File: src/pages/en/guides/deploy/github.md
Comment (sarah11918) [inline src/pages/en/guides/deploy/github.md:18]:
```suggestion
```js title="astro.config.mjs"
```

File: src/pages/en/guides/deploy/github.md
Comment (sarah11918) [inline src/pages/en/guides/deploy/github.md:23]:
```suggestion
      site: 'https://astronaut.github.io',
      base: '/my-repo',
```

**Comment by sarah11918** [comment]
@kevinzunigacuellar @swift502 I took the liberty of throwing in an `astro.config.mjs` example. Please check and see what you think! (It only took me 5 commits, so... 😅 )

File: src/pages/en/guides/deploy/github.md
Comment (kevinzunigacuellar) [inline src/pages/en/guides/deploy/github.md:30]: I am trying to sound a bit more friendly here 😇
```suggestion
      You don't need to set a `base` if:
```

File: src/pages/en/guides/deploy/github.md
Comment (kevinzunigacuellar) [inline src/pages/en/guides/deploy/github.md:18]: highlight base and site options
```suggestion
    ```js title="astro.config.mjs" ins={4-5}
```

File: src/pages/en/guides/deploy/github.md
Comment (sarah11918) [reply src/pages/en/guides/deploy/github.md:30]: Haha, sometimes you don't need to be friendly -- you need to make sure they DON'T DO something! I would honestly keep as is!

File: src/pages/en/guides/deploy/github.md
Comment (sarah11918) [reply src/pages/en/guides/deploy/github.md:18]: I debated this myself, but sure, let's do it!

**Comment by sarah11918** [comment]
It's Friday night, and I think this is good to go! If we spot anything else later, let's just make a new PR.  😅 

Thanks SO MUCH y'all! @kevinzunigacuellar @swift502 @fredoist @Yan-Thomas!  GO TEAM!  🥳

**Pull request merged**
